### PR TITLE
add service_discovery system-probe module

### PR DESCRIPTION
**Pull request by rarguelloF**
### What does this PR do?

This PR adds a new `service_discovery` module to system-probe. This module is enabled by default (same as the `service_discovery` corecheck), linux only and exposes 2 endpoints:
- `/service_discovery/open_ports` -> returns the open ports in the current host.
- `/service_discovery/procs/{pid}` -> returns information about the proc with the given pid.

### Motivation

This functionality was in the `service_discovery` check itself, but it requires root level permissions, so it has been moved to system-probe.

### CODE REVIEW

File: pkg/collector/corechecks/servicediscovery/model/model.go
Comment (jonbodner) [inline pkg/collector/corechecks/servicediscovery/model/model.go:7]: It's not idiomatic Go to have a separate model package. I think we might do that in other places in the agent, unfortunately.

File: pkg/collector/corechecks/servicediscovery/service_detector.go
Comment (jonbodner) [inline pkg/collector/corechecks/servicediscovery/service_detector.go:55]: might want to make this its own function so you can test it separately

File: pkg/collector/corechecks/servicediscovery/model/model.go
Comment (rarguelloF) [reply pkg/collector/corechecks/servicediscovery/model/model.go:7]: Yes, we do in other places of the agent, but we don't necessarily need to do it here. Would it be better to be moved to the `servicediscovery` package?

Comment (rarguelloF) [reply pkg/collector/corechecks/servicediscovery/model/model.go]: Moved to `servicediscovery` package.

Comment (rarguelloF) [reply pkg/collector/corechecks/servicediscovery/model/model.go]: ended up adding it back because of import cycle 😢

File: cmd/system-probe/config/adjust.go
Comment (brycekahle) [inline cmd/system-probe/config/adjust.go:50]: Rather than this, you will want to declare the config value in `pkg/config/setup/system_probe.go` in `InitSystemProbeConfig`. You can set the default there.

**Review by urseberry — APPROVED** [review]

File: pkg/process/net/common.go
Comment (brycekahle) [inline pkg/process/net/common.go:385]: I'd suggest following the pattern of the other module URLs which are defined in `common_linux.go` and `common_windows.go` since the host part of the URL is platform-dependent.

File: cmd/system-probe/config/adjust.go
Comment (brycekahle) [inline cmd/system-probe/config/adjust.go:50]: I would suggest doing this also though, so the core agent has an accurate understanding if the module is supposed to be enabled or not.
```suggestion
    if config.IsContainerized() {
        cfg.Set(serviceDiscoveryNS("enabled"), false, model.SourceAgentRuntime)
    }
```

File: cmd/system-probe/modules/service_discovery_linux.go
Comment (brycekahle) [inline cmd/system-probe/modules/service_discovery_linux.go:43]: If you most this to `adjust.go` as suggested, I think you can remove it here.

File: cmd/system-probe/modules/service_discovery_linux.go
Comment (brycekahle) [inline cmd/system-probe/modules/service_discovery_linux.go:61]: I suggest moving this type to a new package `pkg/servicediscovery` (or whatever path you think is appropriate) that has `CODEOWNERS` set to your team, so that code reviews are requested properly. It will also make running tests in Kernel Matrix Testing (KMT) a bit easier.

**Review by brycekahle — CHANGES_REQUESTED** [review]
Some minor changes requested. If you want to run your tests across a bunch of different Linux kernels, let me know. We can get that setup in this PR fairly easily.

File: pkg/collector/corechecks/servicediscovery/systemprobe/module_linux.go
Comment (brycekahle) [inline pkg/collector/corechecks/servicediscovery/systemprobe/module_linux.go:38]: Apologies for the confusion. I think this type `ServiceDiscoveryModule` should stay in `cmd/system-probe/modules`. The reason is just to reduce confusion as to the possible modules in system-probe. Having them all defined in the same package is helpful.

The type `serviceDiscovery` type can stay in this package though. I may also suggest creating a constructor init function for `serviceDiscovery`(for example `func NewServiceDiscoveryModule() (module.Module, error)`) so that imports like `portlist` (since that is internal to `serviceDiscovery`) can remain in this package.

Happy to chat if you disagree/have questions.

Comment (rarguelloF) [reply pkg/collector/corechecks/servicediscovery/systemprobe/module_linux.go]: no problem! 
what you propose makes sense, created `NewServiceDiscoveryModule` in this package.

**Review by hmahmood — APPROVED** [review]

File: pkg/collector/corechecks/servicediscovery/impl_linux.go
Comment (jonbodner) [inline pkg/collector/corechecks/servicediscovery/impl_linux.go:262]: We should make sure that name is almost always returned from service detection and that it is an error condition when that happens.

File: pkg/collector/corechecks/servicediscovery/service_detector.go
Comment (jonbodner) [inline pkg/collector/corechecks/servicediscovery/service_detector.go:81]: could eliminate `hasPWD` entirely:

```
for _, v := range env {
    if strings.HasPrefix(v, "PWD=") {
        return env
    }
}
env = append(env, fmt.Sprintf("PWD=%s",cwd))
return env
```

File: pkg/collector/corechecks/servicediscovery/servicediscovery.go
Comment (jonbodner) [inline pkg/collector/corechecks/servicediscovery/servicediscovery.go:130]: This isn't an error condition, I think. I would expect we would just exit and maybe log at debug level.

File: pkg/config/setup/system_probe.go
Comment (jonbodner) [inline pkg/config/setup/system_probe.go:385]: ah ok, I guess this is the place where we set the default to enabled. Should the container check be moved here? Also, we should document this for when service discovery for windows is added in Q3.

File: pkg/process/net/common.go
Comment (jonbodner) [inline pkg/process/net/common.go:384]: probably don't need the `url` variable

File: pkg/process/net/common.go
Comment (jonbodner) [inline pkg/process/net/common.go:385]: http.MethodGet instead of "GET"

File: pkg/process/net/common_linux.go
Comment (jonbodner) [inline pkg/process/net/common_linux.go:18]: maybe abstract `unix` and `localhost:3333` into `hostPort` in the net/common_linux and net/common_windows, and put the other constants in net/common, defined using the value of `hostPort`

**Review by jonbodner — CHANGES_REQUESTED** [review]

File: pkg/collector/corechecks/servicediscovery/model/model.go
Comment (richl-dd) [inline pkg/collector/corechecks/servicediscovery/model/model.go:32]: What is the benefit of having 2 endpoints (`open_ports` and `/procs/{pid}`) as opposed a single endpoint that returns:
- `PID`, `ProcessName`, `Port`, `Proto`, `Environ`, and `CWD`

Hosting 2 endpoints in the system probe, querying them from core agent, and then needing to combine their responses seems more complex than having a single endpoint that returns everything needed.

File: pkg/process/net/common_linux.go
Comment (brycekahle) [reply pkg/process/net/common_linux.go:18]: That is probably out of scope for this PR. I can create a ticket for my team to tackle this refactor.

File: pkg/collector/corechecks/servicediscovery/usm/nodejs.go
Comment (amarziali) [reply pkg/collector/corechecks/servicediscovery/usm/nodejs.go:29]: @jonbodner there is not what happen in the node.js tracer. I think we should be consistent. There are also many entry point app files that are called `index.js` or `server.js` so I'm not sure about the fact that will be beneficial. 
If we decide to incorporate it we should also have a plan to change the default on the js tracer

File: pkg/collector/corechecks/servicediscovery/servicediscovery.go
Comment (rarguelloF) [reply pkg/collector/corechecks/servicediscovery/servicediscovery.go:130]: Seeing how other checks are done, this `Configure` function is supposed to be used like this (e.g. return an error when the check is disabled from config).

In theory, users won't see this error because we make sure the check is not added to the list to initialize under this condition, but if for some reason they get here, they will just see 1 warn log.

File: pkg/config/setup/system_probe.go
Comment (rarguelloF) [reply pkg/config/setup/system_probe.go:385]: Please check @brycekahle previous review comments, I've followed the approach he recommended.
Makes sense, I can prepare some docs with all the necessary changes when working on the Windows implementation.

File: pkg/collector/corechecks/servicediscovery/service_detector.go
Comment (rarguelloF) [reply pkg/collector/corechecks/servicediscovery/service_detector.go:81]: true, thanks for pointing this out! I just copied the previous implementation which wasn't in its own function and I missed this :D

File: pkg/collector/corechecks/servicediscovery/usm/nodejs.go
Comment (rarguelloF) [reply pkg/collector/corechecks/servicediscovery/usm/nodejs.go:29]: > There are also many entry point app files that are called index.js or server.js so I'm not sure about the fact that will be beneficial.

Should we check the folder name when the script is called `index.js`? I think we are doing this in some other languages right?

File: pkg/collector/corechecks/servicediscovery/impl_linux.go
Comment (rarguelloF) [reply pkg/collector/corechecks/servicediscovery/impl_linux.go:262]: That would change the existing api for the `usm` package to return `result, error` instead of `result, bool`. 
I suggest leaving this refactor for a future PR to keep the scope as small as possible for this one.

File: omnibus/config/software/datadog-agent-finalize.rb
Comment (paulcacheux) [inline omnibus/config/software/datadog-agent-finalize.rb:103]: Probably missing some `#{install_dir}` here

Comment (chouquette) [reply omnibus/config/software/datadog-agent-finalize.rb:103]: Not necessarily  since the file is copied just before in that folder, however this will override any custom config the customer had there.
You can't conditionally copy a file to the user's folders from the omnibus recipes, you'll have to use the package scripts in `omnibus/package-scripts/`

Comment (rarguelloF) [reply omnibus/config/software/datadog-agent-finalize.rb:103]: Thanks for the info! I see a few scripts inside `omnibus/package-scripts/`, do you know which one is the most appropriate for this?

Comment (spencergilbert) [reply omnibus/config/software/datadog-agent-finalize.rb:103]: Thanks for jumping in @chouquette, I didn't think about that aspect yesterday 🙇.

@rarguelloF I'd use the post install scripts to check if the probe config exists, and copy the example over if it doesn't.

File: omnibus/package-scripts/agent-deb/postinst
Comment (brycekahle) [reply omnibus/package-scripts/agent-deb/postinst:113]: All I know is that our documentation tells users to create the file with `0640` permissions, and you will need write permissions to make any changes. The revert PR does not specify the reasons but I doubt they had to do with the system-probe.yaml file because there were a lot of other changes.

**Review by hush-hush — APPROVED** [review]
Looks good for the files own by ASC

**Comment by jonbodner** [comment]
This work in on hold. Closing this PR.

**Pull request closed without merging**
